oak/server/dev/dev_oak_app_loader_main.cc

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2018 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <csignal>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "absl/synchronization/notification.h"
+#include "absl/time/time.h"
+#include "asylo/util/logging.h"
+#include "dev_oak_loader.h"
+#include "include/grpcpp/server.h"
+#include "include/grpcpp/server_builder.h"
+
+ABSL_FLAG(std::string, config, "", "Application configuration file");
+
+void sigint_handler(int) {
+  LOG(QFATAL) << "SIGINT received";
+  exit(1);
+}
+
+int main(int argc, char* argv[]) {
+  absl::ParseCommandLine(argc, argv);
+
+  // We install an explicit SIGINT handler, as for some reason the default one
+  // does not seem to work.
+  std::signal(SIGINT, sigint_handler);
+
+  // Create loader instance.
+  std::unique_ptr<oak::DevOakLoader> loader = absl::make_unique<oak::DevOakLoader>();
+
+  // Load application configuration.
+  std::unique_ptr<oak::ApplicationConfiguration> application_config =
+      oak::ReadConfigFromFile(absl::GetFlag(FLAGS_config));
+
+  asylo::StatusOr<oak::CreateApplicationResponse> result =
+      loader->CreateApplication(*application_config);
+  if (!result.ok()) {
+    LOG(QFATAL) << "Failed to create application";
+    return 1;
+  }
+  oak::CreateApplicationResponse response = result.ValueOrDie();
+  std::stringstream address;
+  address << "0.0.0.0:" << response.grpc_port();
+  std::string application_id(response.application_id());
+  LOG(INFO) << "Oak Application id=" << application_id << ": " << address.str();
+
+  // Wait.
+  absl::Notification server_timeout;
+  server_timeout.WaitForNotificationWithTimeout(absl::Hours(24));
```

Comment:
Is this just a `sleep(86400)` in disguise, or is it doing something more subtle?

Reply:
No, it also can be stopped with a notification:
https://github.com/abseil/abseil-cpp/blob/08a7e7bf972c8451855a5022f2faf3d3655db015/absl/synchronization/notification.h#L90-L93

Reply:
Yes, didn't find notifications either.
Does it mean, that the same code in the `asylo/` is out of date and should be changed to a regular `sleep`?
https://github.com/project-oak/oak/blob/1528e03883bf23e727f858e1342f335ddf041b06/oak/server/asylo/asylo_oak_main.cc#L66-L68

Reply:
I guess so; maybe look at the history and ask whoever wrote that in the first place.

(Here's hoping that wasn't me.)

Reply:
So basically, this function waits either for the server's configured lifetime to expire or for a shutdown of RPC.
The code was taken from the Asylo example (but this was on the v0.4.1):
https://github.com/google/asylo-examples/blob/be6fd99e8fa0a393eaa95d182190e7ee54ccc02b/grpc_server/grpc_server_enclave.cc#L145-L151

Reply:
In the example, the notification object is passed to the gRPC server, which presumably can notify it; can / should we do the same?

Reply:
I'm not sure if we need this logic in the single app server, since it should support multiple clients without stopping by a notification from one of them.

Reply:
So perhaps mention in a comment that this is just a "sleep in disguise", as @daviddrysdale put it? I don't know if there is a more idiomatic way of doing a `sleep` anyways, in a way that relies on `absl::Hours` or similar readable value (as opposed to just `86400`).

Reply:
Done